Allow for nbt predicates

SirGavith opened the pull request:
Allow for nbt predicates
Also allows for textures to start with `./` and the folder to be called either `optifine` or `mcpatcher`
My auto formatter also had a blast.

SirGavith committed:
Allow for nbt predicates

Allow for nbt predicates
Also allows for textures to start with `./` and the folder to be called either `optifine` or `mcpatcher`
My auto formatter also had a blast.

emilyploszaj reviewed (changes requested):
Couple things, biggest is there are too many style deviations in this PR, please format comments like the rest of the code base (`// Comment after space with capitalized first word`), I also think you may have overcommented in some places. Your formatter also messed with a couple things like ranges and made some function chaining look weird as well as tabs before logging statements, I'd prefer you pick those out and return them to how they used to look. Bit nitpicky but you also have a debug log commented out.

Functionality wise I'm a bit concerned with your implementation of nbt matching. This system will work for basic nbt but will silently fail for basically every complex implementation of nbt matching, like patterns or regex, and doesn't understand ordinal array matching to error either. You can see a lot of these examples in the cit format spec. I'd prefer that chime converter recognizes when these features are attempted and errors. Another thing, nbt can be layered in cit (and in chime) but here, only the last defined nbt match will be counted, you'll need to keep track of all of the predicates and merge them into one tidy json. Final critique, it doesn't seem clear to me that the json this thing emits will be clean (generally, this program emits readable json like that in minecraft sources and expected in resource packs.

SirGavith committed:
Stylistic changes

Stylistic changes

SirGavith committed:
More stylistic improvements

SirGavith committed:
Even more stylistic changes

SirGavith committed:
style

SirGavith committed:
style v5

SirGavith committed:
Better NBT predicates and more

Added support for `texture.foo`; It just ignores everything else (Optifine why)
Added support for `damage` predicate
Added specific error for `weight` predicate so it doesn't abort conversion
Added support for custom models (half done)
Added support for multiple nbt predicates
Added support for `pattern`, `ipattern`, `regex`, and `iregex`
Added debug message for generating an override
Fixed bugs

SirGavith committed:
Implement Custom Models

Implemented Custom Models
I also added a few new functions

SirGavith committed:
Implement weight

turns out its only a few lines

SirGavith committed:
Fix ipattern / iregex and fix skulls

SirGavith committed:
Added handheld items and added custom model + texture support

Added support for custom model + texture (why optifine)
Made any vanilla item `.*(_pickaxe|_axe|_sword|_shovel)` have all its overrides use the handheld model

SirGavith committed:
Hoe fix

Uses handheld model for hoes too

SirGavith committed:
Preserve file structure

Preserve file structure
Also did a little prep for multi-texture overrides (bows, rods)

SirGavith committed:
Add sanitization for paths and fix colons in regex

SirGavith committed:
Refactor

SirGavith committed:
Fix paths and minor refactor

Fix paths
Prepare for bow / rod / armor soonTM

SirGavith committed:
Fix bows and rods

Finally fix bows and rods. Thanks Lucy!
Also minor refactor

SirGavith committed:
Armor :eyes:

Armor works
I also fixed the last commit
I also added auto pack.mcmeta pack_version changing for armor to work
Minor refactor too

SirGavith committed:
Fix `player_head`s
